bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/LedgerDirsMonitor.java

```diff
@@ -0,0 +1,191 @@
+        super("LedgerDirsMonitorThread");
+        this.interval = conf.getDiskCheckInterval();
+        this.conf = conf;
+        this.diskChecker = diskChecker;
+        this.diskUsages = ldm.getDiskUsages();
+        this.ldm = ldm;
+    }
+
+    @Override
+    public void run() {
+        while (true) {
+            try {
+                List<File> writableDirs = ldm.getWritableLedgerDirs();
+                // Check all writable dirs disk space usage.
+                for (File dir : writableDirs) {
+                    try {
+                        diskUsages.put(dir, diskChecker.checkDir(dir));
+                    } catch (DiskErrorException e) {
+                        LOG.error("Ledger directory {} failed on disk checking : ", dir, e);
+                        // Notify disk failure to all listeners
+                        for (LedgerDirsListener listener : ldm.getListeners()) {
+                            listener.diskFailed(dir);
+                        }
+                    } catch (DiskWarnThresholdException e) {
+                        LOG.warn("Ledger directory {} is almost full.", dir);
+                        diskUsages.put(dir, e.getUsage());
+                        for (LedgerDirsListener listener : ldm.getListeners()) {
+                            listener.diskAlmostFull(dir);
+                        }
+                    } catch (DiskOutOfSpaceException e) {
+                        LOG.error("Ledger directory {} is out-of-space.", dir);
+                        diskUsages.put(dir, e.getUsage());
+                        // Notify disk full to all listeners
+                        ldm.addToFilledDirs(dir);
+                    }
+                }
+                // Let's get NoWritableLedgerDirException without waiting for the next iteration
+                // in case we are out of writable dirs
+                // otherwise for the duration of {interval} we end up in the state where 
+                // bookie cannot get writable dir but considered to be writable 
+                ldm.getWritableLedgerDirs();
+            } catch (NoWritableLedgerDirException e) {
+                for (LedgerDirsListener listener : ldm.getListeners()) {
+                    listener.allDisksFull();
+                }
+            }
+
+            List<File> fullfilledDirs = new ArrayList<File>(ldm.getFullFilledLedgerDirs());
+            boolean hasWritableLedgerDirs = ldm.hasWritableLedgerDirs();
+            float totalDiskUsage = 0;
+
+            // When bookie is in READONLY mode .i.e there are no writableLedgerDirs:
+            // - Check if the total disk usage is below DiskLowWaterMarkUsageThreshold.
+            // - If So, walk through the entire list of fullfilledDirs and add them back to writableLedgerDirs list if
+            // their usage is < conf.getDiskUsageThreshold.
+            if (hasWritableLedgerDirs || (totalDiskUsage = diskChecker.getTotalDiskUsage(ldm.getAllLedgerDirs())) < conf
```

Comment:
yes, it is the core piece for this change.

@dlg99  (Andrey) tried to explain in the earlier comments, why he decoupled LedgerDirsMonitor from LedgerDirsManager.